### PR TITLE
more dGrid changes

### PR DESCRIPTION
**Pull request by mesalu**
This PR contains 3 changes

- Editor properties are reapplied when its shown, hopefully preserving intended look and feel with out discarding relevant changes. I consulted @EdLeafe  about this a while ago and he didn't present a better solution at the time.

- a quick fix to correct row shading behavior should `dGrid.AlternateRowColoring` be set. Its an unideal fix, relying on further GridCellAttr cloning, but I don't think it'll present further complications. 🤷‍♂ 

- It also adds a property to dColumns to allow clients to specify sorting behavior via a sort key method. This will shift the burden of sorting to programmers who know the contents of the column, rather than library code that can only guess. 
Should sorting fail, a new type error is raised to inform users how to resolve the issue with out seeking the dabo dev team for assistance.

### CODE REVIEW

File: dabo/ui/dGrid.py [hunk 1]
```diff
@@ -4347,7 +4390,6 @@ def __onWxGridCellEditorKillFocus(self, evt):
         self.HideCellEditControl()
         evt.Skip()
 
-
```
Comment (EdLeafe) [inline dabo/ui/dGrid.py]: Please leave the two-line separation between methods.

**Comment by EdLeafe** [comment]
Looks great!

**Comment by jfabiani** [comment]
Nice work the code appears to fix the AlternateRowColoring.  Thanks.

However, some of my grids (created in a dynamic way - my search grid) 
are apparently setting the "dataType" to " '' " or blank in  the 
dGrid->processSort (my editor shows it at 3044)
if dataType is None:
and I change it to
if dataType is None or dataType == '':

This will force the data item 'f' to be evaluated.

I wonder if this is acceptable????

Johnf

**Comment by jfabiani** [comment]
Please disregard my last posting.  I started looking at the code and 
wondered why all that code was there - there had to be a better way.  I 
also realized 'dataType' also did not match python 3 types.    My 
testing seems to work with floats, int, str, boolean, date,and datetime 
(not sure what other types would be used on a grid). I'm really not sure 
if there are any other types to consider. And the older code did not 
resolve any other types. So I removed most of the code and just used the 
following replacement - please review and let me know your thoughts.

#if dataType is None:  - line 3044
                 f = sortList[0][0]
                 if f is None:
                     # We are just poking around, trying to glean the 
datatype, which is prone
                     # to error. The record we just checked is None, so 
try the last record and
                     # then give up.
                     f = sortList[-1][0]

                 sortingStrings = isinstance(f, str)


                 if colObj.SortKey:
                     sortKey = colObj.SortKey
                 elif sortingStrings and not caseSensitive: # Begin 
guessing at sort keys.
                     sortKey = caseInsensitiveSortKey
                 #elif dataType in ("date", "datetime", 'boolean'):
                 elif isinstance(f, datetime.date) or isinstance(f, 
bool) or isinstance(f,bool):
                     # can't compare NoneType to these types:
                     sortKey = noneSortKey
                 else:
                     sortKey = None

                 try:
                     sortList.sort(key=sortKey, reverse=(sortOrder == 
"DESC"))
                 except TypeError:
                     s = "Unable to sort this column. Conflicting, or 
unknown, data types within the column make it impossible to sort 
implicitly. "\
                         "Please supply a sort key method to the column, 
via the SortKey property, to delegate sort behavior to."
                     if colObj.SortKey is not None:
                         # the preceding exception information will be 
useful for developers to debug their SortKey method
                         raise TypeError(s)
                     else:
                         # There is no user provided SortKey
                         raise TypeError(s) from None

**Pull request merged**